### ISSUE TITLE
creating user features (Liberty runtime extensions) - poss doc from wasdev?

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/docs#2950]
https://github.ibm.com/was-WASdev/WASdevnet-website/tree/master/developer.ibm.com/wasdev/docs/create_your_own_simple_hello_world_user_feature

and

```
<p class="wasdev-excerpt">This article is a continuation from the previous "Create your own simple "Hello World" User Feature".</p>

<p dir="ltr">This article is a continuation from the previous "Create your own simple "Hello World" User Feature" which can be found <a href="/wasdev/docs/create_your_own_simple_hello_world_user_feature">here</a>.</p>

<h2>Updating the User Feature Bundle</h2>

<p dir="ltr">Open the "UserFeatureBundle" user feature again and perform the following actions:</p>

<div dir="ltr">
<ol>
    <li>Right click the "src" folder and select <b>New &gt; Interface</b>.</li>
    <li>Give it the following properties:
<ul>
    <li>Package: "com.ibm.wasdev.tutorial.service".</li>
    <li>Name: "ServiceProvider".</li>
</ul>
</li>
    <li>Click finish.</li>
    <li>Create the method:
[code]
				String getHelloWorld();[/code]</li>
    <li>Double click the manifest file under <b>BundleContent &gt; META-INF</b>.</li>
    <li>Click the "Runtime" tab.</li>
    <li>Under "Exported Packages" click add, and select the "com.ibm.wasdev.tutorial.service" package.</li>
</ol>
Next, open the previously created "HelloWorld" class which looks like:

[code]
package com.ibm.wasdev.tutorial.userfeature;
public class HelloWorld {
	public String getHelloWorld() {
		return "Hello World!";
	}
}
			[/code]
<p dir="ltr">and change it to implement the new Interface:</p>

[code]
package com.ibm.wasdev.tutorial.userfeature;
import com.ibm.wasdev.tutorial.service.ServiceProvider;
public class HelloWorld implements ServiceProvider{
	public String getHelloWorld() {
		return "Hello World!";
	}
}
			[/code]
<p dir="ltr">Next we need to make the Service available, to do this we are going to use <b>blueprint</b>:</p>

<ol>
    <li>Right Click the "UserFeatureBundle" bundle and select <b>New &gt; blueprint file</b></li>
    <li>Click Finish.</li>
    <li>Double click the newly created <b>blueprint</b> file under <b>BundleContent &gt; OSGI-INF &gt; blueprint</b>.</li>
    <li>Select the "Design" tab</li>
    <li>Under "Overview" click "Add..."</li>
    <li>Select "Service" and click "OK".</li>
    <li>Under "Service Interface" click "Browse..." and point it to the interface class.</li>
    <li>Under "Service ID" type <b>OSGIServiceRef</b>.</li>
    <li>Under "Bean Reference" click "New...", then "Browse..." in the new window and finally point it to the "HelloWorld" class.</li>
    <li>Click "OK" twice.</li>
</ol>

<h2>Updating the User Feature</h2>
<p dir="ltr">Now we need to update the user feature manifest to include <b>IBM-API-Package</b> and <b>IBM-API-Service</b>, to do this open the previously created Liberty Feature "UserFeature" and double click "Manifest: UserFeature" and add the following to the bottom of the file:</p>

[code]
	IBM-API-Package: com.ibm.wasdev.tutorial.service
	IBM-API-Service: com.ibm.wasdev.tutorial.service.ServiceProvider [/code]
<p dir="ltr">The final step for the User Feature is to now re-install it, to do this right click "UserFeature" and click "Update Feature...".</p>

<h2>Creating a Web Application Bundle to consume our service</h2>
We've now created and registered an OSGi service in a User Feature, lets use it from a servlet:
<ol>
    <li>Click file and select <b>New &gt; OSGi Bundle Project</b>.</li>
    <li>Set the following properties:</li>
</ol>
<ul>
    <li>Name: "HelloWorldWAB".</li>
    <li>Target runtime: "WebSphere Application Server V8.5 Liberty Profile".</li>
    <li>Add Web Support: checked and set to "Web 3.0".</li>
    <li>Click finish.</li>
    <li>Open the manifest on the new project and go to the "Dependencies" tab.</li>
    <li>Under "Imported Packages" delete all of the packages except "javax.servlet", "javax.servlet.annotation" and "javax.servlet.http"</li>
    <li>Click "Add..." by the imported packages and add "com.ibm.wasdev.tutorial.service"</li>
    <li>Click "Add..." by the imported packages and add "javax.naming"</li>
</ul>
<p dir="ltr">Next, we need to create a blueprint file to reference the OSGi Service thats been made available in the User Feature.</p>

<ol>
    <li>Right Click the "HelloWorldWAB" and select <b>New &gt; blueprint file</b></li>
    <li>Click Finish.</li>
    <li>Double click the newly created <b>blueprint</b> file under <b>BundleContent &gt; OSGI-INF &gt; blueprint</b>.</li>
    <li>Select the "Design" tab</li>
    <li>Under "Overview" click "Add..."</li>
    <li>Select "Reference" and click "OK".</li>
    <li>Under "Reference Interface" click "Browse..." and point it to the interface class.</li>
    <li>Under "Reference ID" type <b>OSGIServiceRef</b>.</li>
    <li>Click "OK" twice.</li>
</ol>
<p dir="ltr">Finally, we need to create a servlet to consume the User Feature's service</p>

<ol>
    <li>Right click on the project and select <b>New &gt; Servlet</b></li>
    <li>Set the properties:
<ul>
    <li>Java package: "com.ibm.wasdev.tutorial.servlet"</li>
    <li>Class name: "HelloWorldServlet"</li>
</ul>
</li>
    <li>Click finish.</li>
    <li>Add the following code to the class:
[code]
@WebServlet("/HelloWorldServlet")
public class HelloWorldServlet extends HttpServlet {
	private static final long serialVersionUID = 1L;

	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
		try {
		      InitialContext ic = new InitialContext();
		      ServiceProvider serviceProvider = (ServiceProvider) ic.lookup("blueprint:comp/OSGIServiceRef");
		      response.getWriter().println(serviceProvider.getHelloWorld());
		    } catch (NamingException e) {
		      e.printStackTrace();
		    }
	}
}
					[/code]</li>
</ol>
<h2>Running the application</h2>
<p dir="ltr">Our application is now ready to print the date in a browser! If you had the application installed into a Liberty Profile server then as you made changes it should of been updated and new features installed. If not drag the application project onto the server in the "Servers" view. The "wab-1.0" and "jndi-1.0" should now be enabled automatically in the feature manager of the server configuration. To run the servlet right click on the HelloWorldServlet class and select "Run as..." -&gt; "Run on server". Next time we will look at how to make writing servlets even simpler by using blueprint to define and manage the service dependencies.</p>

</div>
```

and

```
<p class="wasdev-excerpt">This tutorial will take you through creating, deploying and testing a new feature in the WebSphere Application Server Liberty Profile V8.5.Next Alpha. As an example we will create a new console command to query the default JNDI namespace. This can only be done as a feature, not as a Java EE or OSGi application.</p>

by Joe Chacko

<p style="color: red; font-weight: bold">This article is currently being updated as it's a bit out-of-date (some of the steps are done automatically by the tools now)... Feel free to ask questions on <a href="http://stackoverflow.com/questions/tagged/websphere-liberty">StackOverflow</a> or on <a href="https://www.ibm.com/developerworks/community/forums/html/forum?id=11111111-0000-0000-0000-000000002666&ps=25">our forum</a>.</p>

<h2>Prerequisites</h2>

Before you begin please make sure you have followed the <a href="entry/download_wlp_v85next_alpha">instructions to install the WebSphere Application Server Liberty Profile V8.5.Next Alpha and the WebSphere Developer Tools For Eclipse V.Next</a>

<h2>Create a runtime environment and a new server</h2>

Follow these steps:

<ol>
    <li>Select <b>File</b> &gt; <b>New</b> &gt; <b>Other</b></li>
    <li>In the wizard dialog, type this filter text: <b>server</b></li>
    <li>Select the <b>Server</b> wizard and click <b>Next</b></li>
    <li>Choose the server type: <b>WebSphere Application Server V8.5 Liberty Profile</b></li>
    <li>Set the server name: <b>defaultServer</b></li>
    <li>Click <b>Next</b></li>
    <li>Under the installation folder, enter the location where you installed the Liberty Profile runtime environment</li>
    <li>Make sure you are using a Java 7 JRE (otherwise you will need to tweak the example code)</li>
    <li>Click <b>Finish</b></li>
</ol>

You can try starting and stopping this server to test it and to get familiar with the console output. It isn't configured to do anything so it should start and stop quickly with very little console output.

Tip: you may want to reposition your console pane in Eclipse so that you can see it at the same time as the Servers pane. By default they occupy the same location and you have to switch between them.

<h2>Create an OSGi Bundle</h2>

This bundle will contain the code for the feature.

<ol>
    <li><b>File</b> &gt; <b>New</b> &gt; <b>OSGI Bundle Project</b></li>
    <li>Name the bundle with a name like a java package: <b>com.acme.jndi.explorer</b></li>
    <li>Set the target runtime as the Liberty Profile runtime environment you created</li>
    <li>Do NOT add the bundle to an application (uncheck this option)</li>
    <li>Click <b>Next</b></li>
    <li>Click <b>Next</b></li>
    <li>Check the option to generate an Activator</li>
    <li>Click <b>Finish</b></li>
</ol>

<h2>Implement the Bundle</h2>

In the newly created project create a class called <b>com.acme.jndi.explorer.Command</b>, with the following code:

[code]
package com.acme.jndi.explorer;

import java.util.Dictionary;
import java.util.Hashtable;
import java.util.Properties;

import javax.naming.Binding;
import javax.naming.CompoundName;
import javax.naming.Context;
import javax.naming.InitialContext;
import javax.naming.Name;
import javax.naming.NameNotFoundException;
import javax.naming.NamingEnumeration;
import javax.naming.NamingException;

public class Command {
  static final Dictionary PROPERTIES = new Hashtable();
  static {
    // define the command scope as "jndi:"
    PROPERTIES.put("osgi.command.scope", "jndi");
    // define the command names (which map to methods)
    PROPERTIES.put("osgi.command.function", "ls cd".split(" "));
  }
  private static Properties JNDI_PROPS = new Properties();
  static {
    JNDI_PROPS.put("jndi.syntax.direction", "left_to_right");
    JNDI_PROPS.put("jndi.syntax.separator", "/");
  }

  private InitialContext initialContext;
  private Context context;
  private Name name;

  private Context getContext() throws NamingException {
    if (context == null) {
      context = initialContext = new InitialContext();
      name = new CompoundName("", JNDI_PROPS);
    }
    return context;
  }

  /** list the bindings in the current context */
  public void ls() {
    try {
      NamingEnumeration bindings = getContext().listBindings("");
      while (bindings.hasMore()) {
        Binding b = bindings.next();
        System.out.printf(" %s --> %s%n", b.getName(), b.getObject());
      }
    } catch (NamingException e) {
      handleException(e);
    }
  }

  /** change to another context */
  public void cd(String subcontext) {
    try {
      // ensure contexts are initialized
      getContext();
      // deal with special cases first
      switch (subcontext) {
        case ".": return;
        case "..":
          if (context == initialContext) return;
          int len = name.size() - 1;
          name = name.getPrefix(len);
          context = name.isEmpty() ? initialContext : (Context) initialContext.lookup(name);
          return;
      }
      final Context relativeTo;
      boolean isAbsolute = subcontext.startsWith("/");
      if (isAbsolute) {
        relativeTo = initialContext;
        subcontext = subcontext.substring(1);
      } else {
        relativeTo = context;
      }
      Object o = relativeTo.lookup(subcontext);
      if (o instanceof Context) {
        context = (Context) o;
        name = isAbsolute ? new CompoundName(subcontext, JNDI_PROPS) : name.add(subcontext);
      } else {
        System.err.printf("Not a context: %s%n", subcontext);
      }
    } catch (NamingException e) {
      handleException(e, subcontext);
    }
  }

  private void handleException(NamingException ne, String... args) {
    // add catch blocks for each specific type of exception to be handled
    try {
       throw ne;
    } catch (NameNotFoundException e) {
      String name = args[0];
      System.err.printf("Could not find name: %s%n", name);
    } catch (NamingException e) {
      System.err.printf("Caught exception %s%n", e);
      e.printStackTrace();
    }
  }
}
[/code]

Now for the OSGi bit: we need to publish an instance of this class to the service registry. While there are various ways of doing this, we will use the simplest approach: register the service in the bundle activator. Edit the existing Activator class so that it contains the following stop and start code:

[code]
package com.acme.jndi.explorer;
import org.osgi.framework.BundleActivator;
import org.osgi.framework.BundleContext;

public class Activator implements BundleActivator {
  public void start(BundleContext context) throws Exception {
    System.out.println("JNDI explorer bundle starting");
    context.registerService("java.lang.Object", new Command(), Command.PROPERTIES);
}

  public void stop(BundleContext context) throws Exception {
    System.out.println("JNDI explorer bundle stopping");
  }
}
[/code]

Note that we register but never unregister the service. This is because the framework keeps track and unregisters any remaining services automatically when the bundle is deactivated.

Lastly, because our implementation is using the <b>javax.naming</b> package we need to import it explicitly in the bundle manifest:

<ol>
    <li>Double-click on the project's <b>Manifest</b> to open the bundle manifest editor</li>
    <li>Select the <b>Dependencies</b> tab</li>
    <li>Under <b>Imported Packages</b>, click <b>Add</b>, and select <b>javax.naming</b></li>
    <li>Save and close the bundle manifest</li>
</ol>

<h2>Install the bundle</h2>

In order to run the bundle, we need to put it in the right place in our runtime. This is a somewhat manual exercise in the alpha.

Right-click on the <b>com.acme.jndi.explorer</b> project and choose <b>Export</b> &gt; <b>OSGi Bundle or Fragment</b>

<ol>
    <li>Click the <b>Browse</b> button, and navigate to the <b>usr</b> folder of your installed runtime</li>
    <li>Create an <b>extension</b> folder and navigate into it</li>
    <li>Create a <b>lib</b> folder and navigate into it</li>
    <li>Click <b>Save</b></li>
    <li>Click <b>Finish</b></li>
</ol>

Note that the bundle jar is generated with a timestamp. The Liberty Profile will automatically pick up the newest version of the bundle, so it is ok to generate a new bundle alongside an old one later on.

Check that the bundle is in the correct location by refreshing the server configuration project in your <b>Enterprise Explorer</b> view.

<h2>Create the Feature</h2>

Although we have dropped the code into the server runtime, we have not yet made it accessible. To do that we need to define a feature.

<ol>
    <li><b>File</b>&gt;New&gt;Other</li>
    <li>Type the filter text: <b>feature</b></li>
    <li>Select the <b>Liberty Feature Project</b> wizard</li>
    <li>Set the project name: <span style="font-weight: bold;">Acme JNDI Explorer Feature</span></li>
    <li>Choose the Liberty Profile target runtime</li>
    <li>Click <span style="font-weight: bold;">Next</span></li>
    <li>Ensure that the <span style="font-weight: bold;">com.acme.jndi.explorer_1.0.0</span> bundle is selected</li>
    <li>Click <span style="font-weight: bold;">Finish</span></li>
</ol>

<h2>Install the Feature</h2>

You also need to put the feature into the right location in the runtime.

<ol>
    <li>In the <span style="font-weight: bold;">Enterprise Explorer</span>, expand the server configuration project and create a <span style="font-weight: bold;">features</span> folder under the <span style="font-weight: bold;">extension/lib</span> folder</li>
    <li>In the feature project, expand the <span style="font-weight: bold;">OSGI-INF</span> folder and rename <span style="font-weight: bold;">SUBSYSTEM.MF</span> to <span style="font-weight: bold;">jndiexplorer-1.0.mf</span></li>
    <li>Copy the renamed file, and paste it into the newly created features folder</li>
</ol>

<h2>Configure the server</h2>

<ol>
    <li>In the <span style="font-weight: bold;">Servers</span> view, expand the <span style="font-weight: bold;">defaultServer</span></li>
    <li>Double-click on <span style="font-weight: bold;">Server Configuration</span> to open the server.xml</li>
    <li>In the editor select the <span style="font-weight: bold;">Source</span> tab</li>
    <li>Under the <span style="font-weight: bold;">featureManager</span> element, add three new feature elements for <span style="font-weight: bold;">jndi-1.0</span>, <span style="font-weight: bold;">osgiconsole-1.0</span>, and <span style="font-weight: bold;">usr:jndiexplorer-1.0</span>
[code]
<!-- Enable features -->
  jsp-2.2
  jndi-1.0
osgiconsole-1.0
  usr:jndiexplorer-1.0
[/code]</li>
    <li>Back in the <span style="font-weight: bold;">Design</span> tab, under <span style="font-weight: bold;">Configuration Structure</span> select the <span style="font-weight: bold;">Server Configuration</span> and click the <span style="font-weight: bold;">Add... </span>button</li>
    <li>Type the filter text: <span style="font-weight: bold;">jndi</span></li>
    <li>Click <span style="font-weight: bold;">OK</span> to add a <span style="font-weight: bold;">JNDI Entry</span></li>
    <li>Set the name: <span style="font-weight: bold;">homepage</span></li>
    <li>Set the value: <span style="font-weight: bold;">http://www.acme.com</span></li>
    <li>Add some more of these, including names with slashes within them, e.g.: <span style="font-weight: bold;">math/pi</span></li>
    <li>Check the results in the <span style="font-weight: bold;">Source</span> tab and save the file</li>
</ol>

Note that two of the features will show warnings, and you may see the other features marked as spelling errors. You can safely ignore these.

<h2>Trying the feature</h2>

In order to access the console, you will need to specify a port to use.

<ol>
    <li>Add a new file to your server configuration under <span style="font-weight: bold;">servers/defaultServer</span></li>
    <li>Name the file <span style="font-weight: bold;">bootstrap.properties</span></li>
    <li>Edit the file to add the following line:
[code]osgi.console=5678[/code]</li>
</ol>

<ul>
    <li>Start or restart the server</li>
    <li>Open a command prompt or telnet client and connect to <span style="font-weight: bold;">localhost</span> on port <span style="font-weight: bold;">5678</span></li>
    <li>Try out the new commands: <span style="font-weight: bold;">jndi:ls</span> and <span style="font-weight: bold;">jndi:cd</span>
[code]
osgi&gt; jndi:ls
  homepage --&gt; http://www.acme.com
  math --&gt; math
  osgi&gt; jndi:cd math
  osgi&gt; jndi:ls
   e --&gt; 2.72
   pi --&gt; 3.14
[/code]</li>
</ul>

<h2>Updating the feature</h2>

You may have noticed that the jndi:ls command does not distinguish which entries are contexts. Let's update that and also add a pwd command. We can do this without stopping the server.

<ol>
    <li>Comment out the <span style="font-weight: bold;">usr:jndi-explorer-1.0 </span>feature in <span style="font-weight: bold;">server.xml</span> and save it. You should see the update happen straight away in the Eclipse console.</li>
    <li>On line 20 of the <span style="font-weight: bold;">Command.java</span> file, change the <span style="font-weight: bold;">osgi.command.function</span> property so that it lists an additional command:
[code]
PROPERTIES.put("osgi.command.function", "ls cd pwd".split(" "));
[/code]</li>
</ol>

<ul>
    <li>Add a new method called pwd
[code]
public void pwd() {
  try {
    // ensure we are initialized
    getContext();
    // print the current folder name
    System.out.printf("/%s%n", name);
   } catch (NamingException e) {
    handleException(e);
   }
}
[/code]</li>
    <li>Alter the logic of the ls method to distinguish contexts with a trailing slash
[code]
/** list the bindings in the current context */
public void ls() {
try {
   NamingEnumeration bindings = getContext().listBindings("");
   while (bindings.hasMore()) {
    Binding b = bindings.next();
    Object value = b.getObject();
    if (value instanceof Context) {
      System.out.printf(" %s/%n", b.getName());
    } else {
      System.out.printf(" %s --&gt; %s%n", b.getName(), value);
    }
   }
  } catch (NamingException e) {
   handleException(e);
  }
}[/code]</li>
    <li>Save the new <span style="font-weight: bold;">Command</span> class</li>
    <li>Re-export the bundle from the bundle project to the <span style="font-weight: bold;">usr/extension/lib</span> older</li>
    <li>Uncomment the user feature in <span style="font-weight: bold;">server.xml</span> and save it.</li>
</ul>

You can now try the osgi console again. If you kept the previous session open it will now have the new commands. Since the command bundle has been removed and a new one installed and started, the context will be reset.

[code]
osgi&gt; jndi:ls
  homepage --&gt; http://www.acme.com
  math/
 osgi&gt; jndi:cd math
 osgi&gt; jndi:pwd
  /
 osgi&gt; jndi:cd math
 osgi&gt; jndi:pwd
  /math
[/code]

To recap, we have created a new feature, installed it in a Liberty profile runtime, tried it out, and updated it dynamically. Please have a play with this new function in the alpha and let us know what you think of it.
```

**Comment by dmuelle** [comment]
combining with similar #3324
